src/Sparrow.Server/Allocators.Fragment.cs

```diff
@@ -0,0 +1,272 @@
+        }
+
+        public Pointer Allocate(ref FragmentAllocator<TOptions> allocator, int size)
+        {
+            if (size <= allocator._options.BlockSize)
+            {                
+                if (allocator._used + size > allocator._currentBuffer.Size || !allocator._currentBuffer.IsValid)
+                {
+                    // PERF: This should be considered cold code.
+                    // TODO: Check the code layout.
+                    AllocateSegment(ref allocator, size);
+                }
+
+                Pointer ptr = new Pointer(allocator._ptrCurrent, size);
+                allocator._ptrCurrent += size;
+                allocator._used += size;
+                allocator.TotalAllocated += size;
+                allocator.InUse += size;
+
+                return ptr;
+            }
+
+            // The allocation is too big to fit, the memory will be used afterwards as a segment when released.
+            // PERF: For this kind of allocations we dont care if we hit cold code.
+            Pointer segment = allocator._internalAllocator.Allocate(size);
+            allocator._used += segment.Size;
+            allocator.TotalAllocated += segment.Size;
+            allocator.InUse += segment.Size;
+
+            allocator._allocatedSegments.Add(segment);
+            return segment;
+        }
+
+        /// <summary>
+        /// Comparer for comparing two keys, handling equality as beeing greater
+        /// Use this Comparer e.g. with SortedLists or SortedDictionaries, that don't allow duplicate keys
+        /// </summary>
+        /// <typeparam name="TKey"></typeparam>
+        public class DuplicateKeyComparer<TKey> : IComparer<TKey> where TKey : IComparable
+        {
+            public int Compare(TKey x, TKey y)
+            {
+                int result = y.CompareTo(x);
+
+                if (result == 0)
+                    return 1;   // Handle equality as beeing greater
+               return result;
+            }
+        }
+
+        private void AllocateSegment(ref FragmentAllocator<TOptions> allocator, int size)
+        {
+            int bytesLeft = (int)(allocator._currentBuffer.Size - allocator._used);
+            if (bytesLeft > allocator._options.ReuseBlocksBiggerThan)
+            {
+                // The allocation is bigger than ReuseBlocksBiggerThan and therefore we can leave this one available to be used later.
+                allocator._internalPartialReusedSegments.Add(bytesLeft, new Pointer(allocator._ptrCurrent, bytesLeft));
+            }
+
+            // TODO: Time the impact of using a heap structure instead of a SortedList.
+            // Get the biggest segment on the Heap and if size < the biggest one, we avoid the allocation.
+            Pointer segment;
+            if (allocator._internalPartialReusedSegments.Count > 0 && allocator._internalPartialReusedSegments.Values[0].Size >= size)
```

Comment:
So this goes like this... we insert in biggest to smallest in order (unless I made a mistake on `DuplicateKeyComparer`)... If the biggest chunk is not big enough for the allocation, we just go an get a brand new segment of size BlockSize... remember that if the size is actually bigger than block size, we have a dedicated path to deal with it on `Allocate`

Weirdly enough that is actually the proven way to limit internal fragmentation. The most efficient strategy (while it may sound strange) is to take the biggest chunk instead of the smallest. Someone at the math department of one of the top Israel universities was able to figure that one out (if I am not mixing the location of 2 pretty related papers on fragmentation).

Reply:
You are correct, but let's rename `DuplicateKeyComparer` to `InverseDuplicateKeyComparer` to make it clearer what is going on.

Reply:
Done